### PR TITLE
Add TransactWriteItems

### PR DESCRIPTION
**Pull request by ryanblock**
### We really appreciate your commitment to improving Dynalite

To maintain a high standard of quality in our releases, before merging every pull request we ask that you've completed the following:

- [ ] Forked the repo and created your branch from `main`
- [ ] Made sure tests pass (run `npm it` from the repo root)
- [ ] Expanded test coverage related to your changes:
  - [ ] Added and/or updated unit tests (if appropriate)
  - [ ] Added and/or updated integration tests (if appropriate)
- [ ] Updated relevant documentation
- [ ] Summarized your changes in `changelog.md`
- [ ] Linked to any related issues, PRs, etc. below that may relate to, consume, or necessitate these changes

Please also be sure to completed the CLA (if you haven't already).

Learn more about [contributing to Architect here](https://arc.codes/intro/community).

Thanks again!

### CODE REVIEW

File: test/helpers.js [hunk 1]
```diff
@@ -591,6 +592,24 @@ function assertValidation (target, data, msg, done) {
   })
 }
 
+function assertTransactionCanceled (target, data, msg, done) {
+  request(opts(target, data), function (err, res) {
+    if (err) return done(err)
+    if (typeof res.body !== 'object') {
+      return done(new Error('Not JSON: ' + res.body))
```
Comment (dimaqq) [inline test/helpers.js:599]: super-minor: I'd use a template literal instead

File: validations/transactWriteItems.js [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+          children: {
+            TableName: {
+              type: 'String',
+            },
+            ExpressionAttributeValues: {
+              type: 'Map<AttributeValue>',
+              children: 'AttrStruct<ValueStruct>',
+            },
+            ExpressionAttributeNames: {
+              type: 'Map<java.lang.String>',
```
Comment (dimaqq) [inline validations/transactWriteItems.js:35]: Is a Java reference an artefact of the validation library?
It's odd to see it here.

File: validations/transactWriteItems.js [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
```
Comment (dimaqq) [inline validations/transactWriteItems.js:122]: ditto.

File: validations/transactWriteItems.js [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+}
+
+exports.custom = function (data, store) {
+  var i, request, msg, key
+  for (i = 0; i < data.TransactItems.length; i++) {
+    request = data.TransactItems[i]
+    if (request.Put) {
+      for (key in request.Put.Item) {
+        msg = validations.validateAttributeValue(request.Put.Item[key])
+        if (msg) return msg
```
Comment (dimaqq) [inline validations/transactWriteItems.js:151]: subject to style debate:

```suggestion
        if (msg = validations.validateAttributeValue(request.Put.Item[key]))
          return msg
```

File: validations/transactWriteItems.js [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+        },
+      },
+    },
+  },
+}
+
+exports.custom = function (data, store) {
+  var i, request, msg, key
+  for (i = 0; i < data.TransactItems.length; i++) {
+    request = data.TransactItems[i]
```
Comment (dimaqq) [inline validations/transactWriteItems.js:147]: subject to style debate, or maybe performance?

I'd rather move the variable declaration into the loop:

```suggestion
  for (const request of data.TransactItems) {
    let msg;
```

File: test/transactWriteItem.js [hunk 6]
```diff
@@ -0,0 +1,1281 @@
...
+        'Value null at \'transactItems.1.member.put.item\' failed to satisfy constraint: ' +
+                'Member must not be null',
+      ], done)
+    })
+
+    it('should return ValidationException when writing more than 25 items', function (done) {
+      var requests = [], i
+      for (i = 0; i < 26; i++) {
+        requests.push(i % 2 ? { Delete: { Key: { a: { S: String(i) } } } } : { Put: { Item: { a: { S: String(i) } } } })
+      }
```
Comment (dimaqq) [inline test/transactWriteItem.js:147]: Maybe something like this:

```suggestion
      const requests = new Array(26).fill(null).map(() => ({...}))
```

File: test/transactWriteItem.js [hunk 7]
```diff
@@ -0,0 +1,1281 @@
...
+
+var target = 'TransactWriteItems',
+  request = helpers.request,
+  opts = helpers.opts.bind(null, target),
+  assertType = helpers.assertType.bind(null, target),
+  assertValidation = helpers.assertValidation.bind(null, target),
+  assertTransactionCanceled = helpers.assertTransactionCanceled.bind(null, target),
+  assertNotFound = helpers.assertNotFound.bind(null, target)
+
+describe('transactWriteItem', function () {
```
Comment (dimaqq) [inline test/transactWriteItem.js:12]: This file really needs a review, specifically that correct exceptions are raised in corner cases, in a way that dynalite behaviour matches AWS dynamodb.

I'm flagging it here because GitHub hides this diff by default calling it "too large".

**Review by dimaqq — COMMENTED** [review]
Minor comments mostly

File: validations/transactWriteItems.js (hunk 5 above)
Comment (ryanblock) [reply validations/transactWriteItems.js:147]: Agree

File: validations/transactWriteItems.js (hunk 4 above)
Comment (ryanblock) [reply validations/transactWriteItems.js:151]: Those inline assignments within conditionals are clever but are also imo a bit unusual, and certainly more difficult to debug later. They also blow up `eslint:recommended` on [no-cond-assign](https://eslint.org/docs/latest/rules/no-cond-assign).

**Comment by jpshack-at-palomar** [comment]
@ryanblock What is the must fix here to get this merged? If we are just a few hours of work away I'd be happy pitch in.

**Comment by ryanblock** [comment]
@jpshack-at-palomar please read my comments above; this PR has fundamental issues with how continuation passing works, and it cannot be merged in its current state. As a programmer, I do not feel confident making claims that anything is a few hours of work away, let alone something as complex as this.

I was hoping the author would help us get this PR into shape, but since we haven't heard back my plan has been to hopefully snag a few bits here and there, but probably rewrite. As ever, I'd be happy for someone in the community to start work on a fresh PR.

**Comment by ryanblock** [comment]
Well, it's been >1.5y since I looked at this, but I recall seeing a bunch of callbacky code in here that was incorrectly implemented (i.e. a given callbacks can be called multiple times, etc.). I think anyone who wants to work on this feature should start anew. Given the fact that the original author has not made movement to improve this PR, I think it's probably the best idea to close it and allow the next person to start a fresh one.
